### PR TITLE
feat(ext_messages): add hl_id to ext_messages chunks

### LINKED ISSUE #20310: Make it easier to retrieve highlight attributes during `vim.ui_attach` with `ext_messages`
### Feature already in Vim?

_No response_

### Feature description

When using `ext_messages`, I found no easy way to retrieve highlight attributes. `ext_hlstate` looks promising, but doesn't seem to be available within lua.

Would be great if there would be an API call to retrieve the HL attributes.

Current workaround I found by @notomo:

```lua
local M = {}

function M.setup()
  local ffi = require("ffi")
  local ok, err = pcall(
    ffi.cdef,
    [[typedef int32_t RgbValue;
      typedef struct attr_entry {
        int16_t rgb_ae_attr, cterm_ae_attr;
        RgbValue rgb_fg_color, rgb_bg_color, rgb_sp_color;
        int cterm_fg_color, cterm_bg_color;
        int hl_blend;
      } HlAttrs;
      HlAttrs syn_attr2entry(int attr);]]
  )
  if not ok and not err:find("redefine") then
    error(err)
  end
  M.attr2entry = ffi.C.syn_attr2entry
end

function M.attr2entry(attr_id)
  M.setup()
  return M.attr2entry(attr_id)
end

M.cache = {}

function M.get_hl_group(attr_id)
  return "NoiceAttr" .. tostring(attr_id)
end

function M.get_hl(attr_id)
  if not M.cache[attr_id] then
    local attrs = M.attr2entry(attr_id)
    local hl_group = M.get_hl_group(attr_id)
    vim.api.nvim_set_hl(0, hl_group, {
      fg = attrs.rgb_fg_color,
      bg = attrs.rgb_bg_color,
      sp = attrs.rgb_sp_color,
      bold = bit.band(attrs.rgb_ae_attr, 0x02),
      standout = bit.band(attrs.rgb_ae_attr, 0x0100),
      italic = bit.band(attrs.rgb_ae_attr, 0x04),
      underline = bit.band(attrs.rgb_ae_attr, 0x08),
      undercurl = bit.band(attrs.rgb_ae_attr, 0x10),
      nocombine = bit.band(attrs.rgb_ae_attr, 0x0200),
      reverse = bit.band(attrs.rgb_ae_attr, 0x01),
      blend = attrs.hl_blend ~= -1 and attrs.hl_blend or nil,
    })
    M.cache[attr_id] = hl_group
  end
  return M.cache[attr_id]
end

return M
```

### PR DESCRIPTION
**Pull request by luukvbaal**
**refactor(message): propagate highlight id instead of attrs**

Problem:  Highlight group id is not propagated to the end of the message call
          stack, where ext_messages are emitted.
Solution: Refactor message functions to pass along highlight group id
          instead of attr id.

**feat(ext_messages): add hl_id to ext_messages chunks**

Problem:  Ext_messages chunks only contain the highlight attr id, which
          is not very useful for vim.ui_attach().
Solotion: Add highlight group id to message chunks, which can easily be
          used to highlight text in the TUI through nvim_buf_set_extmark():
          hl_group = hl_id.

Resolve #20310

### CODE REVIEW

File: runtime/doc/ui.txt [hunk 1]
```diff
@@ -197,6 +197,7 @@ the editor.
 
 	The following keys are deprecated:
 
+	undeprecate?
```
Comment (luukvbaal) [inline runtime/doc/ui.txt:200]: Originally deprecated here: https://github.com/neovim/neovim/commit/fa4c2601000df2d792b0de865da3ac8c43ab723f. With this change the original reasoning no longer holds so undeprecate?

File: src/nvim/autocmd.c [hunk 2]
```diff
@@ -219,14 +219,14 @@ static void au_show_for_event(int group, event_T event, const char *pat)
         // show the group name, if it's not the default group
         if (ac->pat->group != AUGROUP_DEFAULT) {
           if (last_group_name == NULL) {
-            msg_puts_attr(get_deleted_augroup(), HL_ATTR(HLF_E));
+            msg_puts_hl(get_deleted_augroup(), HLF_E + 1, false);
```
Comment (luukvbaal) [inline src/nvim/autocmd.c:222]: Adding `+ 1` at each call site is unfortunate. Do we want to adapt the enum and account for it elsewhere?

**Comment by zeertzjq** [comment]
What is the point of this?

**Review by luukvbaal — COMMENTED** [review]
Started as an idea to be able to store message highlights as extmarks. Dropped it because it didn't seem worth the churn. Finished it because @bfredl suggested it would be a welcome change regardless.

Possibly resolves #20310.

**Comment by luukvbaal** [comment]
> What is the point of this

Ran into this while still working on an internal replacement for the message grid. `extmark_set()` expects a highlight group ID, not attributes, so the ID needed to be reserved. See review comment.

Besides that, the point is that the highlight id is more useful than it's attributes for ext_messages consumers(#20310).

**Comment by justinmk** [comment]
> What is the point of this?

A concise problem/solution description in each PR is very helpful. Even though https://github.com/neovim/neovim/issues/27811 also provides useful context, it's very helpful for each PR to say what problem it is solving.

**Comment by luukvbaal** [comment]
@folke could you confirm that this would resolve the need for the workaround you mentioned in #20310?

If so, should I rebase this @zeertzjq @justinmk @bfredl or do you not think it's worth the churn?

**Comment by folke** [comment]
Yes, looks great!

**Comment by bfredl** [comment]
Looks good at a quick glance. Please go ahead and rebase and I will do a final review.

File: src/nvim/autocmd.c (hunk 2 above)
Comment (bfredl) [reply src/nvim/autocmd.c:223]: I'll adjust the enum to start with 1 in a follow up refactor, this is good enough for now.

File: src/nvim/ex_getln_defs.h [hunk 3]
```diff
@@ -49,7 +49,7 @@ struct cmdline_info {
   int cmdfirstc;                ///< ':', '/', '?', '=', '>' or NUL
   int cmdindent;                ///< number of spaces before cmdline
   char *cmdprompt;              ///< message in front of cmdline
-  int cmdattr;                  ///< attributes for prompt
+  int cmdhlid;                  ///< highlight id for prompt
```
Comment (bfredl) [inline src/nvim/ex_getln_defs.h:52]: `cmd_hl_id` (or maybe `cmdhl_id`). it is better to be consistent with always using `hl_id` elsewhere, never `hlid`

File: src/nvim/globals.h [hunk 4]
```diff
@@ -156,7 +156,7 @@ EXTERN bool msg_scrolled_ign INIT( = false);
 EXTERN bool msg_did_scroll INIT( = false);
 
 EXTERN char *keep_msg INIT( = NULL);         // msg to be shown after redraw
-EXTERN int keep_msg_attr INIT( = 0);         // highlight attr for keep_msg
+EXTERN int keep_msg_hlid INIT( = 0);         // highlight id for keep_msg
```
Comment (bfredl) [inline src/nvim/globals.h:159]: `hl_id`

File: src/nvim/statusline.c [hunk 5]
```diff
@@ -582,9 +582,10 @@ void win_redr_ruler(win_T *wp)
 
   if (ui_has(kUIMessages) && !part_of_status) {
     MAXSIZE_TEMP_ARRAY(content, 1);
-    MAXSIZE_TEMP_ARRAY(chunk, 2);
+    MAXSIZE_TEMP_ARRAY(chunk, 3);
     ADD_C(chunk, INTEGER_OBJ(attr));
     ADD_C(chunk, CSTR_AS_OBJ(buffer));
+    ADD_C(chunk, INTEGER_OBJ(0));
```
Comment (bfredl) [inline src/nvim/statusline.c:588]: Hmm, we probably should either calculate the actual hl_id, or just leave this change out completely (we can add it later if it is actually needed)

**Review by bfredl — APPROVED** [review]
otherwise LGTM

File: src/nvim/statusline.c (hunk 5 above)
Comment (luukvbaal) [reply src/nvim/statusline.c]: Right, think I just forgot about this one as a TODO, thanks. In the current logic it is always equal to `HLF_MSG`, added an assert to catch if that ever changes and more logic is needed.

**Pull request merged**